### PR TITLE
vere: adds jets for en/decoding base16

### PR DESCRIPTION
**Pull request by joemfb**
Jets for the new implementations in #3576.

### CODE REVIEW

**Comment by joemfb** [comment]
Once #3612 is merged, I'll update this PR to use the slab abi and only allocate once per call.

File: pkg/urbit/noun/retrieve.c [hunk 1]
```diff
@@ -1107,6 +1107,39 @@ u3r_mp(mpz_t   a_mp,
...
+u3r_short(c3_w  a_w,
+          u3_atom b)
+{
+  c3_assert( u3_none != b );
+  c3_assert( c3y == u3a_is_atom(b) );
+
+  if ( c3y == u3a_is_cat(b) ) {
+    switch ( a_w ) {
+      case 0:  return b & 0xffff;
+      case 1:  return b >> 16;
```
Comment (Fang-) [inline pkg/urbit/noun/retrieve.c:1124]: I'd expect `0` to be no data at all, and `1` to be two bytes (which is what a short is, right?)... so I'm probably misunderstanding what this is supposed to be doing.

**Review by Fang- — COMMENTED** [review]
Looks fine, but there's parts I don't fully get. Probably missing something trivial/unexpected.

File: pkg/urbit/noun/retrieve.c (hunk 1 above)
Comment (joemfb) [reply pkg/urbit/noun/retrieve.c:1124]: `a_w` is a zero-based index into the atom, considered as an array of shorts (ie, uint16). In the case of a direct atom, `u3r_short(0, a)` is the 16 least-significant bits, and `u3r_short(1, a)` is the 16 most significant bits.

This is the same interface as `u3r_bytes()` and `u3r_words()`, modulo the difference in size.

File: pkg/urbit/jets/e/base.c [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+  {
+    c3_w byt_w = u3r_met(3, inp);
+
+    if ( byt_w & 1 ) {
+      ret_o = _of_hex_odd(inp, len_w, byt_w, buf_y);
+    }
+    else {
+      ret_o = _of_hex_even(inp, len_w, buf_y);
+    }
```
[the remark this answers is not in the document]
Comment (joemfb) [reply pkg/urbit/jets/e/base.c:133]: The hex-decoded length in bytes is equal to the hex-encoded length in bytes, divided by two, rounding up. Ie, `u3r_met(3, output) == u3r_met(4, input)` -- two bytes of hex-encoded input correspond 1 byte of output. If there are an even number of bytes, we can simply process them two at a time (reading them as shorts), but if there are odd-number of bytes of input, we must specially handle the most significant, and our pairs of input bytes span input shorts.

I'll add a comment.

File: pkg/urbit/noun/retrieve.c (hunk 1 above)
Comment (Fang-) [reply pkg/urbit/noun/retrieve.c:1124]: > his is the same interface as `u3r_bytes()` and `u3r_words()`

`_byte` and `_word`, right?

This makes sense now, in any case. Thanks!

Comment (joemfb) [reply pkg/urbit/noun/retrieve.c:1124]: Right, that's what I meant.

Comment (joemfb) [reply pkg/urbit/noun/retrieve.c:1124]: I thought about adding `u3r_shorts()` as well, but I don't have an immediate use-case, and it raises (new) endian-ness questions.

**Review by Fang- — APPROVED** [review]
Elaboration and changes look good! (:

**Pull request merged**